File: src/main/java/me/wawwior/utils/event/EventFactory.java
package me.wawwior.utils.event;

import java.util.Collections;
import java.util.Set;
import java.util.WeakHashMap;
import java.util.function.Function;

public interface EventFactory {

    Set<ArrayBackedEvent<?>> ARRAY_BACKED_EVENTS = Collections.newSetFromMap(new WeakHashMap<>());

    static <T> Event<T> createArrayBackedEvent(Function<Integer, T[]> arrayFactory, Function<T[], T> invokerFactory) {
        ArrayBackedEvent<T> event = new ArrayBackedEvent<>(arrayFactory, invokerFactory);
        ARRAY_BACKED_EVENTS.add(event);
        return event;
    }

    static <T> Event<T> createArrayBackedEvent(Function<Integer, T[]> arrayFactory, T emptyInvoker, Function<T[], T> invokerFactory) {
        return createArrayBackedEvent(arrayFactory, listeners -> {
            if (listeners.length == 0) {
                return emptyInvoker;
            }
            if (listeners.length == 1) {
                return listeners[0];
            }
            return invokerFactory.apply(listeners);
        });
    }

    static void rebuildAllInvokers() {
        ARRAY_BACKED_EVENTS.forEach(ArrayBackedEvent::buildInvoker);
    }

}

File: src/main/kotlin/me/wawwior/utils/event/EventFactory.kt
package me.wawwior.utils.event

import com.google.common.collect.MapMaker
import me.wawwior.utils.event.implementation.ArrayBackedEvent
import java.util.*

object EventFactory {
    private val ARRAY_BACKED_EVENTS : MutableSet<ArrayBackedEvent<*>> = Collections.newSetFromMap(WeakHashMap())

    @JvmStatic
    fun <T> createArrayBackedEvent(arrayFactory: (Int) -> Array<T>, invokerFactory: (Array<T>) -> T) : Event<T> {
        val event = ArrayBackedEvent(arrayFactory, invokerFactory)
        ARRAY_BACKED_EVENTS.add(event as ArrayBackedEvent<*>)
        return event
    }

    @JvmStatic
    fun <T> createArrayBackedEvent(arrayFactory: (Int) -> Array<T>, emptyInvoker: T, invokerFactory: (Array<T>) -> T) : Event<T> {
        return createArrayBackedEvent(arrayFactory) { listeners ->
            if (listeners.isEmpty()) {
                emptyInvoker
            } else if (listeners.size == 1) {
                listeners[0]
            } else {
                invokerFactory(listeners)
            }
        }
    }

    @JvmStatic
    fun rebuildAllInvokers() {
        ARRAY_BACKED_EVENTS.forEach { it.buildInvoker() }
    }
}